src/riscv/lib/src/machine_state.rs

```diff
@@ -829,133 +828,99 @@ mod tests {
-                .flat_map(u32::to_le_bytes)
-                .collect::<Vec<_>>()
-                .try_into()
-                .unwrap();
-            state
-                .core
-                .main_memory
-                .write_all(start_ram, &instrs)
-                .unwrap();
-
-            // Configure the machine in such a way that the first instruction will override the
-            // second instruction. The second instruction behaves differently.
-            let address_to_write = start_ram + 4;
-            state.core.hart.xregisters.write(t0, address_to_write);
-            state
-                .core
-                .hart
-                .xregisters
-                .write(t1, I_LOAD_6_INTO_T2 as u64);
+            let res =
+                state
+                    .machine_state
+                    .step_max_handle::<()>(Bound::Unbounded, |machine, _exc| {
+                        let syscall_number = machine.core.hart.xregisters.read(a7) as i64;
+
+                        // The `block-cache-tester` kernel will issue a system call with number -1
+                        // to indicate that it has reached the signal point. This is our cue that we
+                        // have produced the right "start" state.
+                        if syscall_number == -1 {
+                            return Err(());
+                        }
+
+                        let shall_continue = handle_system_call(
+                            machine,
+                            &mut state.system_state,
+                            &mut state.status,
+                            &mut state.reveal_request,
+                            StdoutDebugHooks,
+                        );
+                        Ok(shall_continue)
+                    });
+
+            assert_eq!(
+                res.error,
+                Some(()),
+                "Program didn't make it to the signal point"
+            );
 
             state
         };
 
-        // Perform 2 steps consecutively in one backend.
-        let state = {
-            let mut state = LocalMachineState::<F>::bind(
-                <LocalLayout as CloneLayout>::clone_allocated(
-                    base_state.struct_ref::<FnManagerIdent>(),
-                ),
+        let alt_state = {
+            // Clone the base state to get rid of any ephemeral state that was created.
+            let refs = base_state.struct_ref::<FnManagerIdent>();
+            let refs = PvmLayout::<M64M, TestCacheConfig>::clone_allocated(refs);
+            let mut state = Pvm::<M64M, TestCacheConfig, Interpreted<M64M, F>, F>::bind(
+                refs,
                 InterpretedBlockBuilder,
             );
 
-            state.step().unwrap();
-            state.step().unwrap();
+            // We want to run the kernel until it exits as that is a good point to compare.
+            loop {
+                let _steps = state.eval_max(StdoutDebugHooks, Bound::Unbounded);
```

Comment:
Although the pvm state is the real source of truth - and this is what should be tested - I wonder if at some point we would want to create a similar test but for tracking debug output. Or maybe stdout?

Reply:
Yes

Reply:
Not sure. The standard output is empty on success. So what would the regression changes look like?

Only in case of an error would you get a regression change. However, the test fails before and will print the output for debugging.

How do you envision the block cache tester evolving such that a regression capture is useful?